### ISSUE TITLE
Crashes Foound!

### ISSUE THREAD

**Issue by artofpipe** [ranjian0/building_tool#24]
Hi! the tool is amazing! is not perfect geometry for games, but can be cleanned, the most stressfull Glitches I've found until now of all are:

- when Slab Thickness is set to 0 CRASHES Blender in its totality.
- Slab with an offset so it has to be zero out.
- Adding Floors in different situations crashes Blender in its totality. (also turns the top plane into an n-gon for no reason)
- Stairs not generated and crashed Blender if intended to be tweaked.
- Floor plans when Random can not only crash blend but its geometries are n-gons, so no use.

adding windows zero them  out allows to generate the base for a general layout for later to be filled in. same with the door, that can be spaced in any place and even add windows thanks to the planes it creates: all of that requires cleaning due to the vertex and edges created. but I just wish for this tools to get stabilized so they can be used as a rule of thumb when making generic Buildings.

I Rate this Tools 3/5 CRASHES - I have to save the file every time I want to use one of it features and it only serves for a base for working on top. But I look forward to any improvements done to it!

**Comment by ranjian0** [comment]
1. NGONS - I realize ngons are not ideal especially for export to game engines, but they exist because it is alot easier to do geometry generation with them. Extruding floorplans made of a bunch of quads creates unnecessary geometry. as opposed to one ngon. It is also my assumption that triangulation is easy in blender using `Triangulate` modifier or the `Quads to Tris` operator.

2. Unexpected property values - The addon does not currently have checks for preventing users from entering unsupported and/or extremely high/low values. 0 for slab wall is not supported. This issue needs a combination of solutions such as converting some inputs into sliders and using clamps in the code.

- Again zeroing properties is not fully supported for all building elements, hence the unnecessary edges and vertices. 

In summary the ngon issue should be solved by the user e.g when exporting to game engine. 
Supporting zeroed properties is useful, so I will create a new issue for it.

Thanks for the detailed issue and for the support. I would love to get more feedback from you concerning the addon. Also feel free to post screenshots of issues and/or error logs.

Cheers.

**Comment by artofpipe** [comment]
Hi! Thanks for the response!
I thought that the project could be abandoned 😅 

Like I said, geometry can be cleaned, I do understand why then n-gons when creating the pieces; subdivisions and some other calculations are much easier that way. Cleaning is not a problem when the whole thing is done. I realized that while cleaning 😛 

I noticed that **lots of features** were removed due to instability like balconies and stair variations. those looked amazing in the YouTube demo! Looking forward to have them back, take your time, i know programming this things is not easy 😉 

I will keep play testing the tool and let you know any error, I tried to replicate the error I had of the stairs to send you a screenshot of the message I get, but suddenly it worked 🤷‍♂ I wouldn't know what happened (specially since I've haven't closed Blender, I sleep my laptop everyday)

Current Feedback:
- something I just noticed is that when tweaking in the panel, like that _slab size_, or _step with_ for stairs or any other values; the increments are **0.03** instead of  **0.05** that would be much more appreciated when creating modular things.
- Zeroing values would be cool. but that has been discussed.

Possible additions maybe:
- Ceiling lacks the **Ceiling logs** that stick out in some smaller houses. could be a suggestion for future geometries, that could be added, with its parameters like Log count and scale(in relation to the longest side)
- Windows lack the Window Stool, is decorative but can be appreciated since its part of regular architecture.
- Ceiling Gutter could be added to using the front face of the slab of ceiling for a more complete tool. 
- Pillars when selecting an edge could be possible? adding pillars like a reinforcement of a long building by selecting an edge, I imagine the limitation of it, only working in 3 scenarios: Flat surfaces that are 180° for the half of pillar to stick out, 90° for outside corners showing 3/4 of the pillar and 90° for inside corners showing only 1/4 of the pillar. in any other case getting a message like "Wall need to be flat or 90° corner" for some foolish that tries to break the tool 🤣 

I Hope all that helps. If I can help you in any way, like with geometries let me know! 😃

**Comment by ranjian0** [comment]
- Balcony and Stairs are no longer in `master` branch, the code was really messy and making improvements was almost impossible. I am in the process of doing some refactoring and bug fixes to both. I hope to have them back in `master` within the next 2-3 weeks.

- As for property increments, this consists of a larger issue of the addon. There are lots of properties. Having a unified and intuitive behaviour for all of them is quite challenging. Some properties are more sensitive and therefore need smaller increments. This issue needs very careful and deliberate adjustments. **0.03** is the default increment value in Blender which I admit is not ideal.

- There are lots of features I would like to implement before considering the addon `production` ready. Decorative elements such as window stool and gutters will be considered once I have the main elements such as windows and doors working smoothly. 

- Pillars/Columns are actually a main element, but as you have noticed, there are limitations as to how they should be used. Implementing this element is not trivial but I do intend to have it as a feature in the addon.

**Comment by luckykadam** [comment]
Issues related to `Add Floors` are fixed in #27 
@artofpipe Can you confirm this ?

**Comment by artofpipe** [comment]
Nothing to do with **_New Slab Feature that works wonderfully_** but an error I got before with the stairs, but now with a Door or Stairs or window....

Door:
![image](https://user-images.githubusercontent.com/30891922/76457822-1adf1980-639f-11ea-9d04-c60815c3c9e5.png)

Stairs:
![image](https://user-images.githubusercontent.com/30891922/76458695-8b3a6a80-63a0-11ea-8726-45eafe0d5e21.png)

Window:
![image](https://user-images.githubusercontent.com/30891922/76458815-bcb33600-63a0-11ea-9c20-b2f6b263f1e7.png)

**Apparently it occurs if the face you're adding the door/stairs/window is not a quad.** Hope the report helps :)

**Comment by ranjian0** [comment]
@artofpipe Adding doors/stairs/windows to faces that are not quads is not supported. But it should not be throwing errors either. Will create a new issue for this.

**Comment by ranjian0** [comment]
Most issues in this are fixed.
